Automatic handling of migration to EmbeddedObjects does not work

### How frequently does the bug occur?

Always

### Description

[This feature](https://github.com/realm/realm-core/pull/5737) was added to Realm Core in 12.6.0 and the version of Core currently used in Realm Java is 12.13.0, so I would excpect that manual clearing of orphaned objects is not necessary when migrating from referenced objects to embedded objects. Instead, the app crashes at startup with the message `Cannot convert 'Child' to embedded: at least one object has no incoming links and would be deleted.`.

So, in the current version of the app there is
```
public class Parent extends RealmObject {

@PrimaryKey
private long id;

private Child child;

public Parent() {}

}
```
and

```
public class Child extends RealmObject {

private String name;

public Child() {}

}
```

So far, no care has been taken to delete orphaned children from the database when a parent is deleted, which means for most users there are going to be Child objects that are not referenced by a Parent.
We want to change that now and turn Child into an EmbeddedObject, as follows:
```
// Unchanged
public class Parent extends RealmObject {

@PrimaryKey
private long id;

private Child child;

public Parent() {}

}
```
and

```
@RealmClass(embedded = true)
public class Child extends RealmObject {

private String name;

public Child() {}

}
```

With a new migration added:
```
class Migration : RealmMigration {
    override fun migrate(realm: DynamicRealm, oldVersion: Long, newVersion: Long) {
        val schema = realm.schema
        var oldVersionVal = oldVersion

        if (oldVersionVal == 1L) {
            schema.get("Child")?.isEmbedded = true
            oldVersionVal++
        }
}
```
And on app startup:
```
private function startup() {
        Realm.init(this)
        val config = RealmConfiguration.Builder()
                .name(packageName)
                .schemaVersion(2)
                .migration(Migration())
                .build()
        Realm.setDefaultConfiguration(config)
}
```

However, the automatic handling of the backlinks does not work (i.e. automatic deletion of Child objects in this case). The app crashes on startup.
I know from our iOS app that this feature does work in Realm Swift, leading me to believe there is an issue with Realm Java and not with Realm Core.

### Stacktrace & log output

```
java.lang.RuntimeException: Unable to create application my.app.name.App: java.lang.IllegalStateException: Cannot convert 'Child' to embedded: at least one object has no incoming links and would be deleted.
  at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6739)
  at android.app.ActivityThread.access$1500(ActivityThread.java:256)
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2090)
  at android.os.Handler.dispatchMessage(Handler.java:106)
  at android.os.Looper.loopOnce(Looper.java:201)
  at android.os.Looper.loop(Looper.java:288)
  at android.app.ActivityThread.main(ActivityThread.java:7842)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
  Caused by: java.lang.IllegalStateException: Cannot convert 'Child' to embedded: at least one object has no incoming links and would be deleted.
  at io.realm.internal.Table.nativeSetEmbedded(Native Method)
  at io.realm.internal.Table.setEmbedded(Table.java:796)
  at io.realm.RealmObjectSchema.setEmbedded(RealmObjectSchema.java:595)
  at my.app.name.api.Migration.migrate(Migration.kt:109)
  at io.realm.BaseRealm$6.onMigrationNeeded(BaseRealm.java:892)
  at io.realm.internal.OsSharedRealm.runMigrationCallback(OsSharedRealm.java:576)
  at io.realm.internal.OsSharedRealm.nativeGetSharedRealm(Native Method)
  at io.realm.internal.OsSharedRealm.<init>(OsSharedRealm.java:174)
  at io.realm.internal.OsSharedRealm.getInstance(OsSharedRealm.java:259)
  at io.realm.BaseRealm.<init>(BaseRealm.java:142)
  at io.realm.BaseRealm.<init>(BaseRealm.java:109)
  at io.realm.Realm.<init>(Realm.java:161)
  at io.realm.Realm.createInstance(Realm.java:535)
  at io.realm.RealmCache.createInstance(RealmCache.java:508)
  at io.realm.RealmCache.doCreateRealmOrGetFromCache(RealmCache.java:461)
  at io.realm.RealmCache.createRealmOrGetFromCache(RealmCache.java:422)
  at io.realm.Realm.getDefaultInstance(Realm.java:443)
  at my.app.name.App.startup(App.kt:48)
  at my.app.name.App.onCreate(App.kt:32)
  at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1223)
at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6734)
  at android.app.ActivityThread.access$1500(ActivityThread.java:256) 
  at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2090) 
  at android.os.Handler.dispatchMessage(Handler.java:106) 
  at android.os.Looper.loopOnce(Looper.java:201) 
  at android.os.Looper.loop(Looper.java:288) 
  at android.app.ActivityThread.main(ActivityThread.java:7842) 
  at java.lang.reflect.Method.invoke(Native Method) 
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548) 
  at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003) 
```

### Can you reproduce the bug?

Always

### Reproduction Steps

_No response_

### Version

10.13.2-transformer-api

### What Atlas App Services are you using?

Local Database only

### Are you using encryption?

No

### Platform OS and version(s)

Android API 32

### Build environment

Android Studio version: Android Studio Electric Eel | 2022.1.1 Patch 1 Build #AI-221.6008.13.2211.9514443
Android Build Tools version: 34-rc1
Gradle version: 7.5

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,7 +4,7 @@
 * None.
 
 ### Fixed
-* None.
+* Added support for automatic handling of orphan embedded objects after migrating regular object properties to become embedded objects. (Issue [#7769](https://github.com/realm/realm-java/issues/7769)).
 
 ### Compatibility
 * File format: Generates Realms with format v22. Unsynced Realms will be upgraded from Realm Java 2.0 and later. Synced Realms can only be read and upgraded if created with Realm Java v10.0.0-BETA.1.

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -65,6 +65,10 @@ import io.realm.entities.StringOnlyRequired;
 import io.realm.entities.Thread;
 import io.realm.entities.embedded.EmbeddedSimpleChild;
 import io.realm.entities.embedded.EmbeddedSimpleParent;
+import io.realm.entities.migration.HandleBackLinksChild1;
+import io.realm.entities.migration.HandleBackLinksChild2;
+import io.realm.entities.migration.HandleBackLinksParent1;
+import io.realm.entities.migration.HandleBackLinksParent2;
 import io.realm.entities.migration.MigrationClassRenamed;
 import io.realm.entities.migration.MigrationCore6PKStringIndexedByDefault;
 import io.realm.entities.migration.MigrationFieldRenameAndAdd;
@@ -1716,7 +1720,7 @@ public class RealmMigrationTests {
     }
 
     @Test
-    public void migrate_embeddedObject_handleBackLinks() {
+    public void migrate_embeddedObject_deleteOrphans() {
         RealmConfiguration config1 = configFactory.createConfigurationBuilder()
                 .schema(HandleBackLinksParent1.class, HandleBackLinksChild1.class)
                 .schemaVersion(1)
@@ -1728,8 +1732,8 @@ public class RealmMigrationTests {
             HandleBackLinksParent1 parent = new HandleBackLinksParent1();
             parent.child = child;
             HandleBackLinksChild1 orphan = new HandleBackLinksChild1();
-            realm.insert(parent);
-            realm.insert(orphan);
+            realm.insertOrUpdate(parent);
+            realm.insertOrUpdate(orphan);
         });
         long parentCountV1 = realm1.where(HandleBackLinksParent1.class)
                 .count();
@@ -1778,49 +1782,72 @@ public class RealmMigrationTests {
         realm2.close();
     }
 
+    @Test
+    public void migrate_embeddedObject_clonesChildWhenReferencedMoreThanOnce() {
+        RealmConfiguration config1 = configFactory.createConfigurationBuilder()
+                .schema(HandleBackLinksParent1.class, HandleBackLinksChild1.class)
+                .schemaVersion(1)
+                .build();
+
+        Realm realm1 = Realm.getInstance(config1);
+        HandleBackLinksChild1 child = new HandleBackLinksChild1();
+        HandleBackLinksParent1 parent1 = new HandleBackLinksParent1();
+        HandleBackLinksParent1 parent2 = new HandleBackLinksParent1();
+        realm1.executeTransaction(realm -> {
+            // Copy child and set it as children for both parents.
+            // Now the managed child has two parents. This will not be allowed after converting it
+            // to an embedded object and will result into the child object being duplicated so that
+            // both parents can have a reference to a separate and different embedded object.
+            HandleBackLinksChild1 managedChild = realm.copyToRealm(child);
+            parent1.id = 1L;
+            parent1.child = managedChild;
+            realm.insert(parent1);
+            parent2.id = 2L;
+            parent2.child = managedChild;
+            realm.insert(parent2);
+        });
+        long parentCountV1 = realm1.where(HandleBackLinksParent1.class)
+                .count();
+        assertEquals(2L, parentCountV1);
+        long childCountV1 = realm1.where(HandleBackLinksChild1.class)
+                .count();
+        assertEquals(1L, childCountV1);
+        realm1.close();
+
+        RealmMigration migration = (realm, oldVersion, newVersion) -> {
+            RealmSchema schema = realm.getSchema();
+            if (oldVersion == 1L) {
+                RealmObjectSchema parentSchema = schema.get("HandleBackLinksParent1");
+                assertNotNull(parentSchema);
+                RealmObjectSchema childSchema = schema.get("HandleBackLinksChild1");
+                assertNotNull(childSchema);
+                childSchema.setEmbedded(true, true);
+
+                // Rename classes to avoid conflicts with all other tests
+                parentSchema.setClassName("HandleBackLinksParent2");
+                childSchema.setClassName("HandleBackLinksChild2");
+            }
+        };
+
+        // Create schema v2
+        RealmConfiguration config2 = configFactory.createConfigurationBuilder()
+                .schema(HandleBackLinksParent2.class, HandleBackLinksChild2.class)
+                .schemaVersion(2)
+                .migration(migration)
+                .build();
+
+        // After the migration the only child class is embedded and will not be allowed to have two parents
+        // so the child object will be duplicated
+        Realm realm2 = Realm.getInstance(config2);
+        long parentCountV2 = realm2.where(HandleBackLinksParent2.class)
+                .count();
+        assertEquals(2L, parentCountV2);
+        long childCountV2 = realm2.where(HandleBackLinksChild2.class)
+                .count();
+        assertEquals(2L, childCountV2);
+        realm2.close();
+    }
+
     // TODO Add unit tests for default nullability
     // TODO Add unit tests for default Indexing for Primary keys
-}
-
-// Original parent with a regular object as a child
-//@RealmClass(name = "HandleBackLinksParent1")
-class HandleBackLinksParent1 extends RealmObject {
-    @PrimaryKey
-    public long id;
-
-    public HandleBackLinksChild1 child;
-
-    public HandleBackLinksParent1() {}
-}
-
-// Original child as a regular object
-//@RealmClass(name = "HandleBackLinksChild1")
-class HandleBackLinksChild1 extends RealmObject {
-
-    public String name;
-
-    public HandleBackLinksChild1() {}
-
-}
-
-// Parent, now having an embedded object as a child, respecting table names
-//@RealmClass(name = "HandleBackLinksParent2")
-class HandleBackLinksParent2 extends RealmObject {
-    @PrimaryKey
-    public long id;
-
-    public HandleBackLinksChild2 child;
-
-    public HandleBackLinksParent2() {}
-}
-
-// Child, now as an embedded object, respecting table names
-//@RealmClass(embedded = true, name = "HandleBackLinksChild2")
-@RealmClass(embedded = true)
-class HandleBackLinksChild2 extends RealmObject {
-
-    public String name;
-
-    public HandleBackLinksChild2() {}
-
 }

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -37,6 +37,9 @@ import java.util.concurrent.atomic.AtomicBoolean;
 
 import androidx.test.ext.junit.runners.AndroidJUnit4;
 import androidx.test.platform.app.InstrumentationRegistry;
+
+import io.realm.annotations.PrimaryKey;
+import io.realm.annotations.RealmClass;
 import io.realm.entities.AllTypes;
 import io.realm.entities.AnnotationTypes;
 import io.realm.entities.CatOwner;
@@ -1712,6 +1715,101 @@ public class RealmMigrationTests {
         realm = Realm.getInstance(realmConfig);
     }
 
+    @Test
+    public void migrate_embeddedObject_handleBackLinks() {
+        RealmConfiguration config1 = configFactory.createConfigurationBuilder()
+                .schema(BackLinkParent1.class, BackLinkChild1.class)
+                .schemaVersion(1)
+                .build();
+
+        Realm realm1 = Realm.getInstance(config1);
+        realm1.executeTransaction(realm -> {
+            BackLinkChild1 child = new BackLinkChild1();
+            BackLinkParent1 parent = new BackLinkParent1();
+            parent.child = child;
+            BackLinkChild1 orphan = new BackLinkChild1();
+            realm.insert(parent);
+            realm.insert(orphan);
+        });
+        long parentCount1 = realm1.where(BackLinkParent1.class)
+                .count();
+        assertEquals(1L, parentCount1);
+        long childCount1 = realm1.where(BackLinkChild1.class)
+                .count();
+        assertEquals(2L, childCount1);
+        realm1.close();
+
+        RealmMigration migration = (realm, oldVersion, newVersion) -> {
+            RealmSchema schema = realm.getSchema();
+            if (oldVersion == 1L) {
+                RealmObjectSchema child = schema.get("Child");
+                assertNotNull(child);
+                try {
+                    child.setEmbedded(true);
+                } catch (Exception e) {
+                    assertEquals(IllegalStateException.class, e.getClass());
+                    assertEquals("Cannot convert 'Child' to embedded: at least one object has no incoming links and would be deleted.", e.getMessage());
+                }
+                child.setEmbedded(true, true);
+            }
+        };
+
+        // Create schema v2
+        RealmConfiguration config2 = configFactory.createConfigurationBuilder()
+                .schema(BackLinkParent2.class, BackLinkChild2.class)
+                .schemaVersion(2)
+                .migration(migration)
+                .build();
+
+        // The orphan child is erased
+        Realm realm2 = Realm.getInstance(config2);
+        long parentCount2 = realm2.where(BackLinkParent2.class)
+                .count();
+        assertEquals(1L, parentCount2);
+        long childCount2 = realm2.where(BackLinkChild2.class)
+                .count();
+        assertEquals(1L, childCount2);
+        realm2.close();
+    }
+
     // TODO Add unit tests for default nullability
     // TODO Add unit tests for default Indexing for Primary keys
+}
+
+@RealmClass(name = "Parent")
+class BackLinkParent1 extends RealmObject {
+    @PrimaryKey
+    public long id;
+
+    public BackLinkChild1 child;
+
+    public BackLinkParent1() {}
+}
+
+@RealmClass(name = "Child")
+class BackLinkChild1 extends RealmObject {
+
+    public String name;
+
+    public BackLinkChild1() {}
+
+}
+
+@RealmClass(name = "Parent")
+class BackLinkParent2 extends RealmObject {
+    @PrimaryKey
+    public long id;
+
+    public BackLinkChild2 child;
+
+    public BackLinkParent2() {}
+}
+
+@RealmClass(embedded = true, name = "Child")
+class BackLinkChild2 extends RealmObject {
+
+    public String name;
+
+    public BackLinkChild2() {}
+
 }

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -1777,7 +1777,7 @@ public class RealmMigrationTests {
 }
 
 // Original parent with a regular object as a child
-@RealmClass(name = "Parent")
+@RealmClass(name = "HandleBackLinksParent")
 class BackLinkParent1 extends RealmObject {
     @PrimaryKey
     public long id;
@@ -1788,7 +1788,7 @@ class BackLinkParent1 extends RealmObject {
 }
 
 // Original child as a regular object
-@RealmClass(name = "Child")
+@RealmClass(name = "HandleBackLinksChild")
 class BackLinkChild1 extends RealmObject {
 
     public String name;
@@ -1798,7 +1798,7 @@ class BackLinkChild1 extends RealmObject {
 }
 
 // Parent, now having an embedded object as a child, respecting table names
-@RealmClass(name = "Parent")
+@RealmClass(name = "HandleBackLinksParent")
 class BackLinkParent2 extends RealmObject {
     @PrimaryKey
     public long id;
@@ -1809,7 +1809,7 @@ class BackLinkParent2 extends RealmObject {
 }
 
 // Child, now as an embedded object, respecting table names
-@RealmClass(embedded = true, name = "Child")
+@RealmClass(embedded = true, name = "HandleBackLinksChild")
 class BackLinkChild2 extends RealmObject {
 
     public String name;

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -1718,57 +1718,63 @@ public class RealmMigrationTests {
     @Test
     public void migrate_embeddedObject_handleBackLinks() {
         RealmConfiguration config1 = configFactory.createConfigurationBuilder()
-                .schema(BackLinkParent1.class, BackLinkChild1.class)
+                .schema(HandleBackLinksParent1.class, HandleBackLinksChild1.class)
                 .schemaVersion(1)
                 .build();
 
         Realm realm1 = Realm.getInstance(config1);
         realm1.executeTransaction(realm -> {
-            BackLinkChild1 child = new BackLinkChild1();
-            BackLinkParent1 parent = new BackLinkParent1();
+            HandleBackLinksChild1 child = new HandleBackLinksChild1();
+            HandleBackLinksParent1 parent = new HandleBackLinksParent1();
             parent.child = child;
-            BackLinkChild1 orphan = new BackLinkChild1();
+            HandleBackLinksChild1 orphan = new HandleBackLinksChild1();
             realm.insert(parent);
             realm.insert(orphan);
         });
-        long parentCount1 = realm1.where(BackLinkParent1.class)
+        long parentCountV1 = realm1.where(HandleBackLinksParent1.class)
                 .count();
-        assertEquals(1L, parentCount1);
-        long childCount1 = realm1.where(BackLinkChild1.class)
+        assertEquals(1L, parentCountV1);
+        long childCountV1 = realm1.where(HandleBackLinksChild1.class)
                 .count();
-        assertEquals(2L, childCount1);
+        assertEquals(2L, childCountV1);
         realm1.close();
 
         RealmMigration migration = (realm, oldVersion, newVersion) -> {
             RealmSchema schema = realm.getSchema();
             if (oldVersion == 1L) {
-                RealmObjectSchema child = schema.get("Child");
+                RealmObjectSchema parent = schema.get("HandleBackLinksParent1");
+                assertNotNull(parent);
+                RealmObjectSchema child = schema.get("HandleBackLinksChild1");
                 assertNotNull(child);
                 try {
                     child.setEmbedded(true);
                 } catch (Exception e) {
                     assertEquals(IllegalStateException.class, e.getClass());
-                    assertEquals("Cannot convert 'Child' to embedded: at least one object has no incoming links and would be deleted.", e.getMessage());
+                    assertEquals("Cannot convert 'HandleBackLinksChild1' to embedded: at least one object has no incoming links and would be deleted.", e.getMessage());
                 }
                 child.setEmbedded(true, true);
+
+                // Rename classes to avoid conflicts with all other tests
+                parent.setClassName("HandleBackLinksParent2");
+                child.setClassName("HandleBackLinksChild2");
             }
         };
 
         // Create schema v2
         RealmConfiguration config2 = configFactory.createConfigurationBuilder()
-                .schema(BackLinkParent2.class, BackLinkChild2.class)
+                .schema(HandleBackLinksParent2.class, HandleBackLinksChild2.class)
                 .schemaVersion(2)
                 .migration(migration)
                 .build();
 
         // The orphan child is erased
         Realm realm2 = Realm.getInstance(config2);
-        long parentCount2 = realm2.where(BackLinkParent2.class)
+        long parentCountV2 = realm2.where(HandleBackLinksParent2.class)
                 .count();
-        assertEquals(1L, parentCount2);
-        long childCount2 = realm2.where(BackLinkChild2.class)
+        assertEquals(1L, parentCountV2);
+        long childCountV2 = realm2.where(HandleBackLinksChild2.class)
                 .count();
-        assertEquals(1L, childCount2);
+        assertEquals(1L, childCountV2);
         realm2.close();
     }
 
@@ -1777,43 +1783,44 @@ public class RealmMigrationTests {
 }
 
 // Original parent with a regular object as a child
-@RealmClass(name = "HandleBackLinksParent")
-class BackLinkParent1 extends RealmObject {
+//@RealmClass(name = "HandleBackLinksParent1")
+class HandleBackLinksParent1 extends RealmObject {
     @PrimaryKey
     public long id;
 
-    public BackLinkChild1 child;
+    public HandleBackLinksChild1 child;
 
-    public BackLinkParent1() {}
+    public HandleBackLinksParent1() {}
 }
 
 // Original child as a regular object
-@RealmClass(name = "HandleBackLinksChild")
-class BackLinkChild1 extends RealmObject {
+//@RealmClass(name = "HandleBackLinksChild1")
+class HandleBackLinksChild1 extends RealmObject {
 
     public String name;
 
-    public BackLinkChild1() {}
+    public HandleBackLinksChild1() {}
 
 }
 
 // Parent, now having an embedded object as a child, respecting table names
-@RealmClass(name = "HandleBackLinksParent")
-class BackLinkParent2 extends RealmObject {
+//@RealmClass(name = "HandleBackLinksParent2")
+class HandleBackLinksParent2 extends RealmObject {
     @PrimaryKey
     public long id;
 
-    public BackLinkChild2 child;
+    public HandleBackLinksChild2 child;
 
-    public BackLinkParent2() {}
+    public HandleBackLinksParent2() {}
 }
 
 // Child, now as an embedded object, respecting table names
-@RealmClass(embedded = true, name = "HandleBackLinksChild")
-class BackLinkChild2 extends RealmObject {
+//@RealmClass(embedded = true, name = "HandleBackLinksChild2")
+@RealmClass(embedded = true)
+class HandleBackLinksChild2 extends RealmObject {
 
     public String name;
 
-    public BackLinkChild2() {}
+    public HandleBackLinksChild2() {}
 
 }

--- a/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java
@@ -1776,6 +1776,7 @@ public class RealmMigrationTests {
     // TODO Add unit tests for default Indexing for Primary keys
 }
 
+// Original parent with a regular object as a child
 @RealmClass(name = "Parent")
 class BackLinkParent1 extends RealmObject {
     @PrimaryKey
@@ -1786,6 +1787,7 @@ class BackLinkParent1 extends RealmObject {
     public BackLinkParent1() {}
 }
 
+// Original child as a regular object
 @RealmClass(name = "Child")
 class BackLinkChild1 extends RealmObject {
 
@@ -1795,6 +1797,7 @@ class BackLinkChild1 extends RealmObject {
 
 }
 
+// Parent, now having an embedded object as a child, respecting table names
 @RealmClass(name = "Parent")
 class BackLinkParent2 extends RealmObject {
     @PrimaryKey
@@ -1805,6 +1808,7 @@ class BackLinkParent2 extends RealmObject {
     public BackLinkParent2() {}
 }
 
+// Child, now as an embedded object, respecting table names
 @RealmClass(embedded = true, name = "Child")
 class BackLinkChild2 extends RealmObject {
 

--- a/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksChild1.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksChild1.java
@@ -1,0 +1,12 @@
+package io.realm.entities.migration;
+
+import io.realm.RealmObject;
+
+// Original child as a regular object
+public class HandleBackLinksChild1 extends RealmObject {
+
+    public String name;
+
+    public HandleBackLinksChild1() {}
+
+}

--- a/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksChild2.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksChild2.java
@@ -1,0 +1,14 @@
+package io.realm.entities.migration;
+
+import io.realm.RealmObject;
+import io.realm.annotations.RealmClass;
+
+// Child, now as an embedded object, respecting table names
+@RealmClass(embedded = true)
+public class HandleBackLinksChild2 extends RealmObject {
+
+    public String name;
+
+    public HandleBackLinksChild2() {}
+
+}

--- a/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksParent1.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksParent1.java
@@ -1,0 +1,14 @@
+package io.realm.entities.migration;
+
+import io.realm.RealmObject;
+import io.realm.annotations.PrimaryKey;
+
+// Original parent with a regular object as a child
+public class HandleBackLinksParent1 extends RealmObject {
+    @PrimaryKey
+    public long id;
+
+    public HandleBackLinksChild1 child;
+
+    public HandleBackLinksParent1() {}
+}

--- a/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksParent2.java
+++ b/realm/realm-library/src/androidTest/java/io/realm/entities/migration/HandleBackLinksParent2.java
@@ -1,0 +1,14 @@
+package io.realm.entities.migration;
+
+import io.realm.RealmObject;
+import io.realm.annotations.PrimaryKey;
+
+// Parent, now having an embedded object as a child, respecting table names
+public class HandleBackLinksParent2 extends RealmObject {
+    @PrimaryKey
+    public long id;
+
+    public HandleBackLinksChild2 child;
+
+    public HandleBackLinksParent2() {}
+}

--- a/realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp
+++ b/realm/realm-library/src/main/cpp/io_realm_internal_Table.cpp
@@ -1120,11 +1120,11 @@ JNIEXPORT jboolean JNICALL Java_io_realm_internal_Table_nativeIsEmbedded(JNIEnv*
     return false;
 }
 
-JNIEXPORT jboolean JNICALL Java_io_realm_internal_Table_nativeSetEmbedded(JNIEnv* env, jclass, jlong j_table_ptr, jboolean j_embedded)
+JNIEXPORT jboolean JNICALL Java_io_realm_internal_Table_nativeSetEmbedded(JNIEnv* env, jclass, jlong j_table_ptr, jboolean j_embedded, jboolean j_handle_backlinks)
 {
     try {
         TableRef table = TableRef(TBL_REF(j_table_ptr));
-        table->set_table_type(to_bool(j_embedded) ? Table::Type::Embedded : Table::Type::TopLevel);
+        table->set_table_type(to_bool(j_embedded) ? Table::Type::Embedded : Table::Type::TopLevel, to_bool(j_handle_backlinks));
         return true;
     }
     CATCH_STD()

--- a/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
@@ -584,7 +584,11 @@ public abstract class RealmObjectSchema {
      *     </li>
      * </ul>
      *
-     * @throws IllegalStateException if the class could not be converted because it broke some of the Embedded Objects invariants.
+     * @param embedded If @{code true}, the class type will be turned into an embedded class, and
+     * must satisfy the constraints defined above. If @{code false}, the class will be turn into
+     * a normal class. An embeded class can always be turned into a non-embedded one.
+     * @throws IllegalStateException if the class could not be converted because it broke some of
+     * the Embedded Objects invariants.
      * @see RealmClass#embedded()
      */
     public void setEmbedded(boolean embedded) {
@@ -592,8 +596,8 @@ public abstract class RealmObjectSchema {
     }
 
     /**
-     * Converts the class to be embedded or not allowing automatic handling of backlinks, including cloning of children objects
-     * with multiple parents.
+     * Converts the class to be embedded or not, while also providing automatic handling of objects
+     * that break some of the constraints for making the class embedded.
      * <p>
      * A class can only be marked as embedded if the following invariants are satisfied:
      * <ul>
@@ -607,19 +611,39 @@ public abstract class RealmObjectSchema {
      *     </li>
      * </ul>
      *
-     * This variant adds on {@link RealmObjectSchema#setEmbedded(boolean)} to add support for automatic handling of unsatisfied
-     * invariants after converting the property into an embedded object property. In addition, children objects with multiple
-     * parents that are converted to embedded objects will be cloned to keep the link between parent and child.
+     * If some of these constraints are broken you can ask Realm to resolve them automatically using
+     * the @{code resolveEmbeddedClassConstraints} parameter. Setting this to @{code true} will
+     * do the following:
+     * <ul>
+     *     <li>
+     *         An object with 0 parents, i.e. no other objects have a reference to it, will be
+     *         deleted.
+     *     </li>
+     *     <li>
+     *         An object with more than 1 parent, i.e. 2 or more objects have a reference to it,
+     *         will be copied so each copy have exactly one parent.
+     *     </li>
+     *     <li>
+     *         Objects with a primary key defined will still throw an IllegalStateException and
+     *         cannot be converted.
+     *     </li>
+     * </ul>
      *
-     * @throws IllegalStateException if the class could not be converted because it broke some of the Embedded Objects invariants.
+     * @param embedded If @{code true}, the class type will be turned into an embedded class, and
+     * must satisfy the constraints defined above. If @{code false}, the class will be turn into
+     * a normal class. An embeded class can always be turned into a non-embedded one.
+     * @param resolveEmbeddedClassConstraints whether or not to automatically fix broken constraints
+     * if @{code embedded} was set to true. See above for a full description of what that entails.
+     * @throws IllegalStateException if the class could not be converted because it broke some of
+     * the Embedded Objects invariants and these could not be resolved automatically.
      * @see RealmClass#embedded()
      */
-    public void setEmbedded(boolean embedded, boolean handleBackLinks) {
+    public void setEmbedded(boolean embedded, boolean resolveEmbeddedClassConstraints) {
         if (hasPrimaryKey()) {
             throw new IllegalStateException("Embedded classes cannot have primary keys. This class " +
                     "has a primary key defined so cannot be marked as embedded: " + getClassName());
         }
-        boolean setEmbedded = table.setEmbedded(embedded, handleBackLinks);
+        boolean setEmbedded = table.setEmbedded(embedded, resolveEmbeddedClassConstraints);
         if (!setEmbedded && embedded) {
             throw new IllegalStateException("The class could not be marked as embedded as some " +
                     "objects of this type break some of the Embedded Objects invariants. In order to convert " +

--- a/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
@@ -588,11 +588,36 @@ public abstract class RealmObjectSchema {
      * @see RealmClass#embedded()
      */
     public void setEmbedded(boolean embedded) {
+        setEmbedded(embedded, false);
+    }
+
+    /**
+     * Converts the class to be embedded or not.
+     * <p>
+     * A class can only be marked as embedded if the following invariants are satisfied:
+     * <ul>
+     *     <li>
+     *         The class is not allowed to have a primary key defined.
+     *     </li>
+     *     <li>
+     *         All existing objects of this type, must have one and exactly one parent object
+     *         already pointing to it. If 0 or more than 1 object has a reference to an object
+     *         about to be marked embedded an {@link IllegalStateException} will be thrown.
+     *     </li>
+     * </ul>
+     *
+     * This variant adds on {@link RealmObjectSchema#setEmbedded(boolean)} to add support for automatic deletion of orphan objects
+     * after converting the property into a embedded object class.
+     *
+     * @throws IllegalStateException if the class could not be converted because it broke some of the Embedded Objects invariants.
+     * @see RealmClass#embedded()
+     */
+    public void setEmbedded(boolean embedded, boolean deleteOrphanChildren) {
         if (hasPrimaryKey()) {
             throw new IllegalStateException("Embedded classes cannot have primary keys. This class " +
                     "has a primary key defined so cannot be marked as embedded: " + getClassName());
         }
-        boolean setEmbedded = table.setEmbedded(embedded);
+        boolean setEmbedded = table.setEmbedded(embedded, deleteOrphanChildren);
         if (!setEmbedded && embedded) {
             throw new IllegalStateException("The class could not be marked as embedded as some " +
                     "objects of this type break some of the Embedded Objects invariants. In order to convert " +

--- a/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
@@ -606,18 +606,18 @@ public abstract class RealmObjectSchema {
      *     </li>
      * </ul>
      *
-     * This variant adds on {@link RealmObjectSchema#setEmbedded(boolean)} to add support for automatic deletion of orphan objects
-     * after converting the property into a embedded object class.
+     * This variant adds on {@link RealmObjectSchema#setEmbedded(boolean)} to add support for automatic handling of unsatisfied
+     * invariants after converting the property into an embedded object property.
      *
      * @throws IllegalStateException if the class could not be converted because it broke some of the Embedded Objects invariants.
      * @see RealmClass#embedded()
      */
-    public void setEmbedded(boolean embedded, boolean deleteOrphanChildren) {
+    public void setEmbedded(boolean embedded, boolean handleBackLinks) {
         if (hasPrimaryKey()) {
             throw new IllegalStateException("Embedded classes cannot have primary keys. This class " +
                     "has a primary key defined so cannot be marked as embedded: " + getClassName());
         }
-        boolean setEmbedded = table.setEmbedded(embedded, deleteOrphanChildren);
+        boolean setEmbedded = table.setEmbedded(embedded, handleBackLinks);
         if (!setEmbedded && embedded) {
             throw new IllegalStateException("The class could not be marked as embedded as some " +
                     "objects of this type break some of the Embedded Objects invariants. In order to convert " +

--- a/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
+++ b/realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java
@@ -592,7 +592,8 @@ public abstract class RealmObjectSchema {
     }
 
     /**
-     * Converts the class to be embedded or not.
+     * Converts the class to be embedded or not allowing automatic handling of backlinks, including cloning of children objects
+     * with multiple parents.
      * <p>
      * A class can only be marked as embedded if the following invariants are satisfied:
      * <ul>
@@ -607,7 +608,8 @@ public abstract class RealmObjectSchema {
      * </ul>
      *
      * This variant adds on {@link RealmObjectSchema#setEmbedded(boolean)} to add support for automatic handling of unsatisfied
-     * invariants after converting the property into an embedded object property.
+     * invariants after converting the property into an embedded object property. In addition, children objects with multiple
+     * parents that are converted to embedded objects will be cloned to keep the link between parent and child.
      *
      * @throws IllegalStateException if the class could not be converted because it broke some of the Embedded Objects invariants.
      * @see RealmClass#embedded()

--- a/realm/realm-library/src/main/java/io/realm/internal/Table.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/Table.java
@@ -967,7 +967,5 @@ public class Table implements NativeObject {
 
     private static native boolean nativeIsEmbedded(long nativeTableRefPtr);
 
-//    private static native boolean nativeSetEmbedded(long nativeTableRefPtr, boolean isEmbedded);
-
     private static native boolean nativeSetEmbedded(long nativeTableRefPtr, boolean isEmbedded, boolean handleBackLinks);
 }

--- a/realm/realm-library/src/main/java/io/realm/internal/Table.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/Table.java
@@ -793,7 +793,17 @@ public class Table implements NativeObject {
      * some invariant was broken when trying to change the state
      */
     public boolean setEmbedded(boolean embedded) {
-        return nativeSetEmbedded(nativeTableRefPtr, embedded);
+        return setEmbedded(embedded, false);
+    }
+
+    /**
+     * Returns true if the state was changed, false if not. If false was returned, it meant
+     * some invariant was broken when trying to change the state. The {@code deleteOrphanChildren}
+     * parameter tells Core to automatically delete objects that become orphan children during
+     * migrations.
+     */
+    public boolean setEmbedded(boolean embedded, boolean deleteOrphanChildren) {
+        return nativeSetEmbedded(nativeTableRefPtr, embedded, deleteOrphanChildren);
     }
 
     @Nullable
@@ -957,5 +967,7 @@ public class Table implements NativeObject {
 
     private static native boolean nativeIsEmbedded(long nativeTableRefPtr);
 
-    private static native boolean nativeSetEmbedded(long nativeTableRefPtr, boolean isEmbedded);
+//    private static native boolean nativeSetEmbedded(long nativeTableRefPtr, boolean isEmbedded);
+
+    private static native boolean nativeSetEmbedded(long nativeTableRefPtr, boolean isEmbedded, boolean handleBackLinks);
 }

--- a/realm/realm-library/src/main/java/io/realm/internal/Table.java
+++ b/realm/realm-library/src/main/java/io/realm/internal/Table.java
@@ -798,12 +798,13 @@ public class Table implements NativeObject {
 
     /**
      * Returns true if the state was changed, false if not. If false was returned, it meant
-     * some invariant was broken when trying to change the state. The {@code deleteOrphanChildren}
-     * parameter tells Core to automatically delete objects that become orphan children during
-     * migrations.
+     * some invariant was broken when trying to change the state. The {@code handleBackLinks}
+     * parameter tells Core to automatically handle all unsatisfied invariants for backlinks, e.g.
+     * children becoming orphan or cloning objects with multiple references during migrations from
+     * a regular object to an embedded object.
      */
-    public boolean setEmbedded(boolean embedded, boolean deleteOrphanChildren) {
-        return nativeSetEmbedded(nativeTableRefPtr, embedded, deleteOrphanChildren);
+    public boolean setEmbedded(boolean embedded, boolean handleBackLinks) {
+        return nativeSetEmbedded(nativeTableRefPtr, embedded, handleBackLinks);
     }
 
     @Nullable

